edap.py

```diff
@@ -213,14 +42,181 @@ def __init__(self, hostname, admin_cn, password, domain=None):
+        modlist = [(ldap.MOD_ADD, "memberUid", [uid.encode("ASCII")])]
+        self.ldap.modify_s(group_fqdn, modlist)
+
+    def make_uid_member_of_division(self, uid, name):
+        group_fqdn = f"cn={name},{self.ldap.DIVISIONS_GROUP}"
+        return self.make_uid_member_of(uid, group_fqdn)
+
+    def make_uid_member_of_service_group(self, uid, name):
+        group_fqdn = f"cn={name},{self.ldap.SERVICES_GROUP}"
+        return self.make_uid_member_of(uid, group_fqdn)
+
+    def remove_uid_member_of(self, uid, group_fqdn):
+        if self.object_exists_at(group_fqdn, "posixGroup") == 0:
+            raise ConstraintError(f"Group {group_fqdn} doesn't exist.")
+        if not self.uid_is_member_of_group(group_fqdn, uid):
+            return
+        if self.user_of_uid_exists(uid) == 0:
+            msg = f"User of uid '{uid}' doesn't exist, so we can't add it to any group."
+            raise ConstraintError(msg)
+        modlist = [(ldap.MOD_DELETE, "memberUid", [uid.encode("ASCII")])]
+        self.ldap.modify_s(group_fqdn, modlist)
+
+    def remove_uid_member_of_division(self, uid, name):
+        group_fqdn = f"cn={name},{self.ldap.DIVISIONS_GROUP}"
+        return self.remove_uid_member_of(uid, group_fqdn)
+
+    def remove_uid_member_of_service_group(self, uid, name):
+        group_fqdn = f"cn={name},{self.ldap.SERVICES_GROUP}"
+        return self.remove_uid_member_of(uid, group_fqdn)
+
+
+class LdapFranchiseMixin(object):
+
+    def create_franchise(self, name):
+        if not self.subobject_exists_at(self.ldap.FRANCHISES, "organizationalUnit"):
+            self.create_org_unit(self.ldap.FRANCHISES, self.ldap.FRANCHISES_GROUP)
+        if not self.subobject_exists_at(f"cn={name},{self.ldap.FRANCHISES}", "posixGroup"):
+            dic = self.create_group_dict(f"{name}")
+            dic["description"] = self.label_franchise(name).encode("UTF-8")
+            self.create_group_from_dict(f"cn={name},{self.ldap.FRANCHISES_GROUP}", dic)
+
+    def label_franchise(self, name):
+        for code, country_name in c.COUNTRIES_CODES.items():
+            if name.startswith(code):
+                return country_name
+        raise KeyError(f"Invalid country code to match '{name}'")
+
+    def create_all_franchises(self, source):
+        for frname in source:
+            self.create_franchise(frname)
+
+
+class LdapDivisionMixin(object):
+
+    def create_division(self, name):
+        if not self.subobject_exists_at(self.ldap.DIVISIONS, "organizationalUnit"):
+            self.create_org_unit(self.ldap.DIVISIONS, self.ldap.DIVISIONS_GROUP)
+        if not self.subobject_exists_at(f"cn={name},{self.ldap.DIVISIONS}", "posixGroup"):
+            dic = self.create_group_dict(f"{name}")
+            self.create_group_from_dict(f"cn={name},{self.ldap.DIVISIONS_GROUP}", dic)
+
+    def create_all_divisions(self, source):
+        for dname in source:
+            self.create_division(dname)
+
+
+def ensure_org_sanity(edap, source):
```

Comment:
Note for later - the function should be able to deduce what does it take to produce a sane organization from it's arguments. It should accept a sequence (list?) of objects in which each one that would represent e.g. a country, a division, a project, ... Then, the function would figure out that all of those objects should have some representation in the LDAP tree, so it will create the corresponding organizational unit, or other LDAP entry.

Reply:
Thanks, looks like that function is not used yet, but will take it into consideration when will work with that one.